Neos.Media/Classes/Domain/Service/AssetVariantGenerator.php

```diff
@@ -93,16 +85,100 @@ public function createVariants(AssetInterface $asset): array
     }
 
     /**
-     * @param Image $originalAsset
+     * @param AssetInterface $asset
+     * @return AssetVariantInterface[] The created variants (if any), with the preset identifier as array key
+     * @throws AssetVariantGeneratorException
+     */
+    public function recreateVariants(AssetInterface $asset): array
+    {
+        // Currently only Image Variants are supported. Other asset classes can be supported, as soon as there is a common
+        // interface for creating and adding variants.
+        if (!$asset instanceof Image) {
+            return [];
+        }
+
+        $createdVariants = [];
+        foreach ($this->getVariantPresets() as $presetIdentifier => $preset) {
+            if ($preset->matchesMediaType($asset->getMediaType())) {
+                foreach ($preset->variants() as $variantIdentifier => $variantConfiguration) {
+                    $createdVariants[$presetIdentifier . ':' . $variantIdentifier] = $this->recreateVariant($asset, $presetIdentifier, $variantIdentifier);
+                }
+            }
+        }
+
+        return $createdVariants;
+    }
+
+    /**
+     * @param AssetInterface $asset
+     * @param string $presetIdentifier
+     * @param string $variantIdentifier
+     * @return AssetVariantInterface The created variant (if any)
```

Comment:
I'm stumbling upon the nullable return type here (and for `recreateVariant()`):
Isn't `null` always something unexpected that the calling side might want to know about?
Since this method can throw an `AssetVariantGeneratorException` anyways I would much prefer a non-nullable return type with a corresponding exception if the variant couldn't be generated, since otherwise there are chances of null pointer errors which can be really hard to debug

Reply:
Good point. I added this to handle non-`Image` somewhat gracefully, but maybe throwing an exception is better. Then again, most of the variant-handling code is "soft" currently, e.g. when asking for a non.-existing variant you get the original back (IIRC).

Reply:
To be honest, I think that's gracefulness applied in a counter-productive way.
If one wants `createVariant()` to ignore any error, a `try/catch` block can be used. But with the nullable return type you _have_ to check for null. And even if you do, you might not know why it didn't work (asset was not an image, variant preset was not a `VariantPreset` (why not?), variantConfiguration was not an instance of `Variant`).
It might make more sense to make the *calling* side more graceful so that it only invokes this method if the asset is an `Image` etc.

> when asking for a non.-existing variant you get the original back

Uh, is that so? I would not expect that. But I get what you say, and maybe it _is_ less confusing to go for one style

Reply:
Well, what could lead to no variant being returned? The obvious case is an error, then an exception makes sense.

But it might well be, that there is no preset (ok, can be seen as an error) or that the preset media type does not match the asset (is that an error?)